harvard-lts/mirador-pdiiif-plugin
# Add initial PDF download plugin

**JIRA Ticket**: [LTSVIEWER-156](https://jira.huit.harvard.edu/browse/LTSVIEWER-156)

[Companion PR for MPS-viewer](https://github.com/harvard-lts/mps-viewer/pull/41)

# What does this Pull Request do?

- Adds initially functionality for PDF download using the [PDIIIF library](https://github.com/jbaiter/pdiiif/tree/main/pdiiif-lib) **for PTO's only**.
- This currently doesn't support MPS manifests ([LTSVIEWER-174](https://jira.huit.harvard.edu/browse/LTSVIEWER-174))
- Currently does not support Firefox or browsers that don't work with the PDIIIF's implementation of the FileSystem API ([LTSVIEWER-172](https://jira.huit.harvard.edu/browse/LTSVIEWER-172))

# How should this be tested?

- On mps-viewer, checkout the companion branch locally which contains the plugin as a dependency
- Restarting the service locally should run NPM install and install the dependencies without any issues. It should also run webpack build which will be essential to see this update too.

- Viewing any object from the demo page you should be able to see the PDF option in the plugin menu (three vertical dots) in Mirador, below the download link.
- Right now, the possible states of this link are:
  - Checking PDF: Menu link is disabled shows text "Checking document" and a spinner icon (as this is currently doubling up as a validity check for MPS manifests it makes sense to appear here, but we may remove this state after [LTSVIEWER-174](https://jira.huit.harvard.edu/browse/LTSVIEWER-174))
  - Something went wrong checking the PDF: Menu link is disabled and shows text "PDF Unavailable"
  - Firefox or browser lacking API support:  Menu link is disabled and shows text "PDF Unavailable"
  - MPS manifest: Menu link is disabled and shows text "PDF Unavailable"
  - Non-PTO (single page object): Menu link is disabled and shows text "PDF Unavailable"
  - IDS PTO manifest: After checking the link is enabled and shows text "PDF download"
- Where PDF download is supported, a modal show display asking the user to confirm they wish to save the current document as a PDF. If a filesize could be estimated, this will be displayed
- A system dialog should present and allow the user to select a destination
- The PDF should appear in that directory automatically

# Test coverage

Yes/No: Are changes in this pull-request covered by:

- unit tests? Yes
- integration tests? No

# Interested parties
@enriquediaz
